test/apps/iodemo/io_demo.cc

```diff
@@ -0,0 +1,447 @@
+    };
+
+    DemoServer(const options_t& test_opts) : P2pDemoCommon(test_opts) {
+    }
+
+    void run() {
+        struct sockaddr_in listen_addr;
+        memset(&listen_addr, 0, sizeof(listen_addr));
+        listen_addr.sin_family      = AF_INET;
+        listen_addr.sin_addr.s_addr = INADDR_ANY;
+        listen_addr.sin_port        = htons(opts().port_num);
+
+        listen((const struct sockaddr*)&listen_addr, sizeof(listen_addr));
+        for (;;) {
+            try {
+                progress();
+            } catch (const std::exception &e) {
+                std::cerr << e.what();
+            }
+        }
+    }
+
+    void handle_io_read_request(UcxConnection* conn, const iomsg_hdr_t *hdr) {
+        // send data
+        conn->send_data(buffer(), opts().data_size, hdr->sn);
+
+        // send response as data
+        IoMessage *response = new IoMessage(opts().iomsg_size, IO_COMP, hdr->sn,
+                                            0);
+        conn->send_data(response->buffer(), opts().iomsg_size, hdr->sn,
+                        response);
+    }
+
+    void handle_io_write_request(UcxConnection* conn, const iomsg_hdr_t *hdr) {
+        conn->recv_data(buffer(), opts().data_size, hdr->sn,
+                        new IoWriteResponseCallback(this, conn, hdr->sn));
+    }
+
+    virtual void dispatch_io_message(UcxConnection* conn, const void *buffer) {
+        const iomsg_hdr_t *hdr = reinterpret_cast<const iomsg_hdr_t*>(buffer);
+
+        if (opts().verbose) {
+            std::cout << "got io message " << io_op_names[hdr->op]
+                      << " sn "   << hdr->sn << " data size " << hdr->data_size
+                      << " conn " << conn    << std::endl;
+        }
+
+        if (hdr->op == IO_READ) {
+            handle_io_read_request(conn, hdr);
+        } else if (hdr->op == IO_WRITE) {
+            handle_io_write_request(conn, hdr);
+        }
+    }
+};
+
+
+class DemoClient : private P2pDemoCommon {
+public:
+    class IoResponseCallback : public UcxCallback {
+    public:
+        IoResponseCallback(long *counter, size_t iomsg_size) :
+            _counter(counter), _buffer(malloc(iomsg_size)) {
+        }
+
+        ~IoResponseCallback() {
+            free(_buffer);
+        }
+
+        virtual void operator()() {
+            ++(*_counter);
+            delete this;
+        }
+
+        void* buffer() {
+            return _buffer;
+        }
+
+    private:
+        long* _counter;
+        void* _buffer;
+    };
+
+    DemoClient(const options_t& test_opts) :
+        P2pDemoCommon(test_opts),
+        _num_sent(0), _num_completed(0) {
+    }
+
+    void do_io_read(UcxConnection *conn, uint32_t sn) {
+        ++_num_sent;
+        send_io_message(conn, IO_READ, sn);
+        conn->recv_data(buffer(), opts().data_size, sn);
+        // receive response as data to ensure ordering, and count completions
+        IoResponseCallback *response = new IoResponseCallback(&_num_completed,
```

Comment:
need to increment num_completed only when both data and response arrived, not only when response arrived